README.md
# tap-fixerio

A [Singer Tap] Tap to extract currency exchange rate data from [fixer.io].

## How to use it

### Install and Run

First, make sure Python 3 is installed on your system or follow these
installation instructions for [Mac] or [Ubuntu].

Then, get an access key from [fixer.io](http://fixer.io). You need at
least the basic plan to get exchange rates from a base currency (such as
USD).

Then, convert `config.sample.json` to
`~/singer.io/tap_fixerio_config.json`; fill out your parameters:

Config Key | Required? | Default | Description
--- | --- | --- | ---
`subscription_plan` | Yes | 'free' | Your fixer.io subscription plan
`start_date` | Yes | Today | The starting date from which rates will be pulled in `YYYY-MM-DD` format (overridden if a state file is passed)
`access_key` | Yes | None | Your fixer.io access key
`symbols` | No | None | An optional list of currency symbols to fetch in the following format: ['USD', 'GBP', etc.]. If not specified fixer will return a list of all exchange rates for each day
`base` | Yes | 'USD' | The base rate to which others will be converted (not available in free subscription plan, for which it is set to 'EUR')

It's recommended to use a virtualenv:

```bash
python3 -m venv ~/.virtualenvs/tap-fixerio
source ~/.virtualenvs/tap-fixerio/bin/activate
pip install -U pip setuptools
pip install -e '.'
```

Like all Singer taps, the output of `tap-fixerio` should be piped to a Singer target. 
Deactivate the `tap-fixerio` virtual environment by using the `deactivate` command, 
then set up the `target-csv` virtual environment according to the instructions
[here](https://github.com/singer-io/target-csv/blob/master/README.md).

Once `target-csv`, or another Singer target, is installed in its own virtual
environment run them with the following command:

```bash
~/.virtualenvs/tap-fixerio/bin/tap-fixerio --config ~/singer.io/tap_fixerio_config.json | ~/.virtualenvs/target-csv/bin/target-csv
```

The data will be written to a file called `exchange_rate.csv` in your
working directory.

```bash
$ cat exchange_rate.csv
AUD,BGN,BRL,CAD,CHF,CNY,CZK,DKK,GBP,HKD,HRK,HUF,IDR,ILS,INR,JPY,KRW,MXN,MYR,NOK,NZD,PHP,PLN,RON,RUB,SEK,SGD,THB,TRY,ZAR,EUR,USD,date
1.3023,1.8435,3.0889,1.3109,1.0038,6.869,25.47,7.0076,0.79652,7.7614,7.0011,290.88,13317.0,3.6988,66.608,112.21,1129.4,19.694,4.4405,8.3292,1.3867,50.198,4.0632,4.2577,58.105,8.9724,1.4037,34.882,3.581,12.915,0.9426,1.0,2017-02-24T00:00:00Z
```

---

Copyright &copy; 2017 Stitch

[Singer Tap]: https://singer.io
[fixer.io]: https://fixer.io
[Mac]: http://docs.python-guide.org/en/latest/starting/install3/osx/
[Ubuntu]: https://www.digitalocean.com/community/tutorials/how-to-install-python-3-and-set-up-a-local-programming-environment-on-ubuntu-16-04


CHANGELOG.md
# Changelog
## 0.1.4
## 0.1.4
  * Dependabot update [#14](https://github.com/singer-io/tap-fixerio/pull/14)

## 0.1.4
  * Dependabot update [#14](https://github.com/singer-io/tap-fixerio/pull/14)

## 0.1.4
  * Dependabot update [#14](https://github.com/singer-io/tap-fixerio/pull/14)

  * Dependabot update [#14](https://github.com/singer-io/tap-fixerio/pull/14)

tap_exchangeratesapi.py
#!/usr/bin/env python3

import json
import sys
import argparse
import time
import requests
import singer
import backoff

from datetime import date, datetime, timedelta

base_url = 'https://api.exchangeratesapi.io/'

logger = singer.get_logger()
session = requests.Session()

DATE_FORMAT='%Y-%m-%d'

def parse_response(r):
    flattened = r['rates']
    flattened[r['base']] = 1.0
    flattened['date'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.strptime(r['date'], DATE_FORMAT))
    return flattened

schema = {'type': 'object',
          'properties':
          {'date': {'type': 'string',
                    'format': 'date-time'}},
          'additionalProperties': True}

def giveup(error):
    logger.error(error.response.text)
    response = error.response
    return not (response.status_code == 429 or
                response.status_code >= 500)

@backoff.on_exception(backoff.constant,
                      (requests.exceptions.RequestException),
                      jitter=backoff.random_jitter,
                      max_tries=5,
                      giveup=giveup,
                      interval=30)
def request(url, params):
    response = requests.get(url=url, params=params)
    response.raise_for_status()
    return response
    
def do_sync(base, start_date):
    logger.info('Replicating exchange rate data from exchangeratesapi.io starting from {}'.format(start_date))
    singer.write_schema('exchange_rate', schema, 'date')

    state = {'start_date': start_date}
    next_date = start_date
    
    try:
        while True:
            response = request(base_url + next_date, {'base': base})
            payload = response.json()

            if datetime.strptime(next_date, DATE_FORMAT) > datetime.utcnow():
                break
            else:
                singer.write_records('exchange_rate', [parse_response(payload)])
                state = {'start_date': next_date}
                next_date = (datetime.strptime(next_date, DATE_FORMAT) + timedelta(days=1)).strftime(DATE_FORMAT)

    except requests.exceptions.RequestException as e:
        logger.fatal('Error on ' + e.request.url +
                     '; received status ' + str(e.response.status_code) +
                     ': ' + e.response.text)
        singer.write_state(state)
        sys.exit(-1)

    singer.write_state(state)
    logger.info('Tap exiting normally')


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-c', '--config', help='Config file', required=False)
    parser.add_argument(
        '-s', '--state', help='State file', required=False)

    args = parser.parse_args()

    if args.config:
        with open(args.config) as file:
            config = json.load(file)
    else:
        config = {}

    if args.state:
        with open(args.state) as file:
            state = json.load(file)
    else:
        state = {}

    start_date = state.get('start_date') or config.get('start_date') or datetime.utcnow().strftime(DATE_FORMAT)
    start_date = singer.utils.strptime_with_tz(start_date).date().strftime(DATE_FORMAT)

    do_sync(config.get('base', 'USD'), start_date)


if __name__ == '__main__':
    main()


tap_fixerio.py
#!/usr/bin/env python3

import json
import sys
import argparse
import time
import requests
import singer
import backoff

from datetime import date, datetime, timedelta

REQUIRED_CONFIG_KEYS = [
    "access_key",
]

base_url = 'https://data.fixer.io/api/'

logger = singer.get_logger()
session = requests.Session()
DATE_FORMAT='%Y-%m-%d'

def parse_response(r):
    flattened = r['rates']
    flattened[r['base']] = 1.0
    flattened['date'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.strptime(r['date'], DATE_FORMAT))
    return flattened

schema = {'type': 'object',
          'properties':
          {'date': {'type': 'string',
                    'format': 'date-time'}},
          'additionalProperties': True}

def giveup(error):
    logger.error(error.response.text)
    response = error.response
    return not (response.status_code == 429 or
                response.status_code >= 500)

@backoff.on_exception(backoff.constant,
                      (requests.exceptions.RequestException),
                      jitter=backoff.random_jitter,
                      max_tries=5,
                      giveup=giveup,
                      interval=30)
def request(url, params):
    response = requests.get(url=url, params=params)
    response.raise_for_status()
    return response

def do_sync(base, start_date, access_key, symbols=None):
    logger.info('Replicating exchange rate data from fixer.io starting from {}'.format(start_date))
    singer.write_schema('exchange_rate', schema, 'date')

    state = {'start_date': start_date}
    next_date = start_date

    try:
        while True:
            if symbols:
                response = request(base_url + next_date, {'base': base, 'access_key': access_key, 'symbols': ','.join(symbols)})
            else:
                response = request(base_url + next_date, {'base': base, 'access_key': access_key})
            payload = response.json()

            if datetime.strptime(next_date, DATE_FORMAT) > datetime.utcnow():
                break
            elif payload.get('error'):
                raise RuntimeError(payload['error'])
            else:
                singer.write_records('exchange_rate', [parse_response(payload)])
                state = {'start_date': next_date}
                next_date = (datetime.strptime(next_date, DATE_FORMAT) + timedelta(days=1)).strftime(DATE_FORMAT)

    except requests.exceptions.RequestException as e:
        logger.fatal('Error on ' + e.request.url +
                     '; received status ' + str(e.response.status_code) +
                     ': ' + e.response.text)
        singer.write_state(state)
        sys.exit(-1)

    singer.write_state(state)
    logger.info('Tap exiting normally')


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-c', '--config', help='Config file', required=False)
    parser.add_argument(
        '-s', '--state', help='State file', required=False)

    #args = parser.parse_args()
    args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)

    if args.config:
        config = args.config
    else:
        config = {}

    if args.state:
        state = args.state
    else:
        state = {}

    start_date = state.get('start_date',
                           config.get('start_date', datetime.utcnow().strftime(DATE_FORMAT)))
    access_key = state.get('access_key', config.get('access_key'))

    do_sync(config.get('base', 'USD'), start_date, access_key, symbols=config.get('symbols', None))


if __name__ == '__main__':
    main()
